### PR TITLE
Admin API

### PR DESCRIPTION
**Pull request by fabxc**
After many shenanigans of how we can implement a full grpc API, I decided to just add handlers for the new admin functionality. It's just a lot to be figured out regarding types, so I'd just defer that to some time later in 2.x

This however gives us snapshots and deletions for the new storage.

### CODE REVIEW

**Comment by brian-brazil** [comment]
This seems generally fine from a quick look.

For #2173 we should get the admin endpoints disabled by a flag by default.

**Comment by Gouthamve** [comment]
Snapshot is a GET request? Shouldn't it be POST?

**Comment by fabxc** [comment]
That probably makes sense, yes.

I just made `[POST] /delete_series` btw. We are not really RESTy as the proper thing would probably be `[DEL] /series` but that would need proto `oneof` field where we can either provide a textual or a typed matcher. In the end we probably don't care about being full REST though but more about having some JSON API at all, right?

**Comment by Gouthamve** [comment]
POST `/delete_series` is okay, but I think we can send a body along with a DELETE request, but not sure if that is a good practice.

**Comment by fabxc** [comment]
The GRPC gateway won't allow us to specify a body for DELETE verbs. Guess just leaving it as post is fine. Changed snapshotting to POST.

File: prompb/rpc.proto [hunk 1]
```diff
@@ -0,0 +1,63 @@
...
+
+option (gogoproto.sizer_all) = true;
+option (gogoproto.marshaler_all) = true;
+option (gogoproto.unmarshaler_all) = true;
+option (gogoproto.goproto_getters_all) = false;
+
+
+service Admin {
+  // Snapshot creates a snapshot of all current data into 'snapshots/<datetime>-<rand>' under
+  // the TSDB's date directory.
```
Comment (brian-brazil) [inline prompb/rpc.proto]: data

File: web/web.go [hunk 2]
```diff
@@ -207,16 +207,20 @@ func New(o *Options) *Handler {
...
 	}
 
-	if o.EnableQuit {
+	if o.EnableLifecycle {
 		router.Post("/-/quit", h.quit)
+		router.Post("/-/reload", h.reload)
+	} else {
+		router.Post("/-/quit", func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(http.StatusForbidden)
+			w.Write([]byte("Lifecycle APIs are not enabled"))
```
Comment (brian-brazil) [inline web/web.go]: At this level, can we mention the flag?

File: web/web.go [hunk 3]
```diff
@@ -207,16 +207,20 @@ func New(o *Options) *Handler {
...
+		router.Post("/-/reload", func(w http.ResponseWriter, _ *http.Request) {
+			w.WriteHeader(http.StatusForbidden)
+			w.Write([]byte("Lifecycle APIs are not enabled"))
+		})
 	}
 
-	router.Post("/-/reload", h.reload)
-	router.Get("/-/reload", func(w http.ResponseWriter, r *http.Request) {
-		w.WriteHeader(http.StatusMethodNotAllowed)
-		fmt.Fprintf(w, "This endpoint requires a POST request.\n")
```
Comment (brian-brazil) [inline web/web.go]: We should keep this, it avoids a fair bit of user confusion. And we should probably add the same for shutdown.

File: web/web.go [hunk 4]
```diff
@@ -48,10 +54,10 @@ import (
 	"github.com/prometheus/prometheus/promql"
 	"github.com/prometheus/prometheus/retrieval"
 	"github.com/prometheus/prometheus/rules"
-	"github.com/prometheus/prometheus/storage"
 	"github.com/prometheus/prometheus/template"
 	"github.com/prometheus/prometheus/util/httputil"
 	api_v1 "github.com/prometheus/prometheus/web/api/v1"
+	apiv2 "github.com/prometheus/prometheus/web/api/v2"
```
Comment (brian-brazil) [inline web/web.go]: this should be api_v2 for consistency with the previous line

File: documentation/dev/api/swagger.json [hunk 5]
```diff
@@ -0,0 +1,123 @@
...
+    "https"
+  ],
+  "consumes": [
+    "application/json"
+  ],
+  "produces": [
+    "application/json"
+  ],
+  "paths": {
+    "/admin/v1/tsdb/delete_series": {
```
Comment (brian-brazil) [inline documentation/dev/api/swagger.json]: I thought this was a v2 api

File: web/web.go (hunk 2 above)
Comment (fabxc) [inline web/web.go]: Even here it is passed in as a configuration option. We don't know that it was set through a flag.
I'd prefer to only do it if it becomes an understanding problem for users.

File: web/web.go (hunk 3 above)
Comment (fabxc) [inline web/web.go]: We do not do the same for POST to the snapshot for example. Where do we draw the line?

File: documentation/dev/api/swagger.json (hunk 5 above)
Comment (fabxc) [inline documentation/dev/api/swagger.json]: Yea, we've to make a call. Either we make it `/v2/{admin,...}` or we declare the admin API as it's own thing, in which case it starts with that as a prefix and is v1 `/admin/v1/...`.

Opinions @Gouthamve @juliusv

File: web/web.go (hunk 3 above)
Comment (brian-brazil) [inline web/web.go]: I'd say just reload/shutdown, as we've seen them cause a fair few user questions which went away when we added this.

File: documentation/dev/api/swagger.json (hunk 5 above)
Comment (Gouthamve) [inline documentation/dev/api/swagger.json]: I don't have a preference, but I think `v2/admin` makes more sense.

Comment (brian-brazil) [inline documentation/dev/api/swagger.json]: I think one unified versioning scheme would be less confusing for users.

Comment (brian-brazil) [inline documentation/dev/api/swagger.json]: I had presumed that a v2 api would include everything, in particular the current querying stuff, so users don't have to worry about which version to use for which features.

Comment (fabxc) [inline documentation/dev/api/swagger.json]: See the PR's initial comment. I started implementing but deferred it for now due to too many open questions.

**Comment by Gouthamve** [comment]
LGTM

**Comment by brian-brazil** [comment]
:+1:

**Pull request merged**